### PR TITLE
chore(vrl): use new `value` crate for type checking in VRL

### PR DESCRIPTION
**Pull request by JeanMertz**
Reverts vectordotdev/vector#11287

### CODE REVIEW

**Comment by JeanMertz** [comment]
I'm still investigating the root-cause of this breaking change.

I've got this minimal reproducible example to pass the tests on this branch, now I'm seeing if it would fail on master:

```
# issue: https://github.com/vectordotdev/vector/pull/11287
# result:
#
# error[E103]: unhandled fallible assignment
#   ┌─ :3:9
#   │
# 3 │ .tags = push(.tags, "bar")
#   │ ------- ^^^^^^^^^^^^^^^^^^
#   │ │       │
#   │ │       this expression is fallible
#   │ │       update the expression to be infallible
#   │ or change this to an infallible assignment:
#   │ .tags, err = push(.tags, "bar")
#   │
#   = see documentation about error handling at https://errors.vrl.dev/#handling
#   = learn more about error code 103 at https://errors.vrl.dev/103
#   = see language documentation at https://vrl.dev

.tags, err = push(.tags, "foo")
.tags = push(.tags, "bar")
```

On first sight, this _should_ not be allowed to compile, since we cannot be sure that `.tags` is an array on the second line (it could be `null` if the first line fails).

**Comment by JeanMertz** [comment]
Verified that this "passes" on master, but then subsequently fails at runtime:

```
issues
expected value at line 1 column 1
  11287 http pipelines blackhole..............................FAILED (expectation)
  "\n\nerror[E103]: unhandled fallible assignment\n  ┌─ :3:9\n  │\n3 │ .tags = push(.tags, \"bar\")\n  │ ------- ^^^^^^^^^^^^^^^^^^\n  │ │       │\n  │ │       this expression is fallible\n  │ │       update the expression to be infallible\n  │ or change this to an infallible assignment:\n  │ .tags, err = push(.tags, \"bar\")\n  │\n  = see documentation about error handling at https://errors.vrl.dev/#handling\n  = learn more about error code 103 at https://errors.vrl.dev/103\n  = see language documentation at https://vrl.dev"
[
  "bar"
]


  Overall result: FAILED

    Number failed: 1
```

**Comment by JeanMertz** [comment]
I'm not sure what changes I made triggered this to be resolved, so I'll have to dig into that, to know the extent to which this addresses an issue in VRL. It might be limited in scope to the `push` function, or it might be a wider problem that is now resolved, but can potentially cause many programs to no longer compile before they deal with the error at compile-time.

**Comment by JeanMertz** [comment]
Actually, I now realize this is _not_ a problem in master, but is a problem in my branch.

I forgot that a fallible assignment does _not_ set the `ok` variant to `null` if the assignment fails, but sets it to the "default value", which in this case means `.tags` would become an empty array if it failed, which would then mean the second line would be infallible because `.tags` is either a populated array, or an empty array, but it would be _an_ array.

I'll investigate further why this regression was introduced, and why it wasn't caught in any of the existing test cases.

**Comment by JeanMertz** [comment]
Verified that the issue has to do with the `default_value`.

In master, the type definition for `push` returns an `array`, in this branch it returns `any`, which then results in the default value becoming `null`, which causes the second expression to fail.

Debugging further to find the actual root cause of the type def returning `any` vs the expected `array`.

File: lib/vrl/compiler/src/type_def.rs [hunk 1]
```diff
@@ -349,8 +349,22 @@ impl TypeDef {
...
 
     pub fn merge(&mut self, other: Self, strategy: merge::Strategy) {
-        self.kind.merge(other.kind, strategy);
         self.fallible |= other.fallible;
+
+        // NOTE: technically we shouldn't do this, but to keep backward compatibility with the
+        // "old" `Kind` implementation, we consider a type that is marked as "any" equal to the old
+        // implementation's `unknown`, which, when merging, would discard that type.
+        //
+        // see: https://github.com/vectordotdev/vector/blob/18415050b60b08197e8135b7659390256995e844/lib/vrl/compiler/src/type_def.rs#L428-L429
```
Comment (pablosichert) [inline lib/vrl/compiler/src/type_def.rs:358]: It's not clear to me which "right" way you envision here or does this comment not imply that we should change any of this logic in the future?

If we intend to change it, I'd suggest opening an issue with the proposed logic and a migration plan.

**Review by pablosichert — APPROVED** [review]
Looks good 👍

We don't expect this to land in `0.20.0`, do we? Given that this PR contains significant complexity and no user-facing change, I'd like to use the next release cycle to iron out any potential issues left.

If we merge, I'd like to make sure that `0.20.0` is cut before this commit. I'm surprised this hasn't been the case already.

File: lib/vrl/compiler/src/type_def.rs (hunk 1 above)
Comment (JeanMertz) [reply lib/vrl/compiler/src/type_def.rs:358]: I'll likely introduce a new merge `Strategy` option that allows ignoring `any` values. It's not "bad" per sé, but it is confusing behaviour without explicitly requesting it as the caller of a type definition merge.

For now, that strategy doesn't exist in the `value` crate, so it lives here in VRL, I'll think on what the best next steps would be to either remove the need for this entirely, or move it to a more "official" place as a strategy in the value crate.

**Pull request merged**